frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/traverse/BaseTraverseAction.java

```diff
@@ -0,0 +1,129 @@
+    val project = getProject(anActionEvent);
+
+    boolean yesNoResult = true;
+    if (branchLayout != null && branchLayout.getRootEntries().nonEmpty() && gitRepository != null) {
+      if (PropertiesComponent.getInstance(project).getBoolean(SHOW_TRAVERSE_INFO, /* defaultValue */ true)) {
+
+        val traverseInfoDialog = new InfoDialog(
+            project,
+            getString("action.GitMachete.BaseTraverseAction.dialog.traverse-approval.title"),
+            getString("action.GitMachete.BaseTraverseAction.dialog.traverse-approval.text.HTML"),
+            SHOW_TRAVERSE_INFO,
+            /* myHeight */ 100);
+
+        yesNoResult = traverseInfoDialog.showAndGet();
+      }
+
+      if (yesNoResult) {
+        val entryName = getNameOfBranchUnderAction(anActionEvent);
+        if (entryName != null) {
+          checkoutAndTraverse(gitRepository, graphTable, entryName);
+        } else {
+          LOG.warn("Skipping traverse action because entryName is undefined");
+        }
+      }
+    }
+  }
+
+  private void checkoutAndTraverse(GitRepository gitRepository, BaseEnhancedGraphTable graphTable, String branchName) {
+    log().debug(() -> "Queuing '${branchName}' branch checkout background task");
+
+    @UI Runnable traverseRunnable = () -> traverse(gitRepository, graphTable, branchName);
+    Runnable repositoryRefreshRunnable = () -> graphTable.queueRepositoryUpdateAndModelRefresh(traverseRunnable);
+    Runnable callInAwtLater = () -> ModalityUiUtil.invokeLaterIfNeeded(NON_MODAL, repositoryRefreshRunnable);
+    val gitBrancher = GitBrancher.getInstance(gitRepository.getProject());
+    val repositories = Collections.singletonList(gitRepository);
+    gitBrancher.checkout(/* reference */ branchName, /* detach */ false, repositories, callInAwtLater);
+  }
+
+  @UIEffect
+  private void traverse(GitRepository gitRepository, BaseEnhancedGraphTable graphTable, String branchName) {
+    val repositorySnapshot = graphTable.getGitMacheteRepositorySnapshot();
+    val branchLayout = repositorySnapshot != null ? repositorySnapshot.getBranchLayout() : null;
+    val gitMacheteBranch = repositorySnapshot != null ? repositorySnapshot.getManagedBranchByName(branchName) : null;
+
+    if (gitMacheteBranch != null) {
+      @UI Runnable traverseNextEntry = () -> {
+        var nextBranch = branchLayout != null ? branchLayout.findNextEntry(branchName) : null;
+        if (nextBranch != null) {
+          checkoutAndTraverse(gitRepository, graphTable, nextBranch.getName());
+        }
+      };
+
+      if (gitMacheteBranch.isNonRoot()) {
+        assert repositorySnapshot != null : "repositorySnapshot is null";
+        new TraverseSyncToParent(gitRepository, graphTable, repositorySnapshot, gitMacheteBranch.asNonRoot(), traverseNextEntry)
+            .sync();
+      } else {
+        new TraverseSyncToRemote(gitRepository, graphTable, gitMacheteBranch, traverseNextEntry).sync();
```

Comment:
Huh I don't like a method called `sync()` that executes mostly **a**synchronously ;D

I mean, ofc _sync_ is unrelated here to _(a)synchronous execution_... but still, a bit misleading (even if only on subconscious level).

Maybe just `execute()`?